V1 innerblocks documentatie (#13)
Studio-Lemon/wp-lemon-docs: src/docs/extend/innerblocks.mdx

---
title: Innerblocks

> description: How to use innerblocks in wp-lemon

Reviewer: ```suggestion
description: How to use InnerBlocks in wp-lemon
```

---

> By using innerblocks, you can nest blocks within other blocks. 

Reviewer: ```suggestion
By using `InnerBlocks`, you can nest blocks within other blocks.
```

This allows you to create more complex and flexible layouts by placing blocks inside parent blocks.

> **Note:**
> A single block can only contain one `InnerBlocks` component.

## Building your template

> When using innerblocks it's common to create a template for the innerblocks. 

Reviewer: ```suggestion
When using `InnerBlocks` it's common to create a template for the `InnerBlocks`. 
```

> This will prefill the InnerBlocks component when inserted and will make it easier to use for you and your clients. 

Reviewer: ```suggestion
This will prefill the `InnerBlocks `component with blocks within the template when inserted and will make it easier to use for you and your clients. 
```

> In wp-lemon we do this inside the `block_context`. 

Reviewer: ```suggestion
In wp-lemon we do this inside the `HighGround\Bulldozer\AbstractBlockRenderer::block_context` method.
```

``` php

	public function block_context($context): array
	{
		$template = [
			[
				'core/heading',
				[

> 					'placeholder' => __('Add a title', 'wp-lemon'),

Reviewer: ```suggestion
					'placeholder' => __('Add a title', 'wp-lemon-child'),
```

					'level'       => 3,
				],
			],
			[
				'core/paragraph',
				[

> 					'placeholder' => __('Add some text', 'wp-lemon'),

Reviewer: ```suggestion
					'placeholder' => __('Add some text', 'wp-lemon-child'),
```

				],
			],
		];

		$args = [
			'InnerBlocks' => self::create_inner_blocks(false, $template),
		];

		return array_merge($context, $args);
	}

```

> In the example above we've used a heading and a paragraph block, 

Reviewer: ```suggestion
In the example above we've used a heading and a paragraph block.
```

> ofcourse you can add other core blocks or even wp-lemon blocks. You can do this like so:

Reviewer: ```suggestion
Ofcourse, you can add other core blocks or even wp-lemon blocks. You can do this like so:
```

``` php

	$template = [
		['acf/content-card'],
];

```

## Template properties 
By adding extra parameters in the `create_inner_blocks` function we can add some extra functionality. 
In the list below you can see each parameter with it's functionality in the right order.

| Parameter     | Function     |
|--------------|--------------|

> | $allowed_blocks | This parameter allows you to make a selection of blocks that are allowed in the innerblocks |

Reviewer: ```suggestion
| $allowed_blocks | This parameter allows you to make a selection of blocks that are allowed in the `InnerBlocks` |
```

| $template | This parameter refers to your innerblocks template  |
| $classes | This parameter can be used to give custom classnames to your innerblocks container |

> | $orientation | ???? |

Reviewer: ```suggestion
| $orientation | Use this parameters to change the orientation of the arrow to move the child blocks left/right or up down. This is useful in column layouts. Can be `horizontal`, `vertical` or `false` |
```

| $templatelock | This parameter gives the ability to lock down the template so no changes can be made |

> **Note:**
> If a parameter is not given it will default to false.

### Allowed blocks
The `$allowed_blocks` parameter allows you to make a selection of blocks that are allowed in the innerblocks. 
We used the `card-grid` block from wp-lemon as an example. 
In this case the we only wanted to allow specific cards we've made. 
We did this like so:
 
``` php
	public function block_context($context): array
	{
		$template = [
			['acf/content-card'],
		];

> 		$allowed_blocks = apply_filters('wp-lemon/filter/block/card-grid/allowed-blocks', ['acf/content-card', 'acf/image-card', 'acf/price-card']);

Reviewer: ```suggestion
		$allowed_blocks = ['acf/content-card', 'acf/image-card', 'acf/price-card'];
```

Reviewer: De filters gebruiken we alleen in wp-lemon zelf zodat je dit vanuit je child kan aanpassen. In het child thema is dat dus niet nodig.

		$args = [
			'InnerBlocks' => self::create_inner_blocks($allowed_blocks, $template),
		];

		return array_merge($context, $args);
	}

```

> In the example above we've added it as a parameter in the `create_inner_blocks` function, but the filter that is used could also be used in `libary/hooks.php`.

Reviewer: ```suggestion
In the example above we've added it as a parameter in the `create_inner_blocks` function. Note: all wp-lemon blocks that use `InnerBlocks` have a filter to expand/tighten the child blocks allowed.
```

> You can find more documentation on this right [here](/extend/hooks#filters-available-php-block-declarations).

Reviewer: Nice!

### Classes
To add custom classes to your innerblocks use the third parameter. You can put all the classnames in a string seperated by a space like so:

``` php

$args = [
    'InnerBlocks' => self::create_inner_blocks($template, 'classname classname'),
];

```

### Orientation

> Ik heb geen flauw idee wat dit doet

Reviewer: ```suggestion
Orientation can be used to display the move icons for the child blocks in a left/right or a top/down orientation. This is useful for block layouts were child blocks are next to each other. In that case you would use `horizontal````

### Templatelock
The fifth and last parameter gives the ability to lock down the template. By default this parameter is false, 

> meaning the template will show in your wordpress editor, but you don't have to stick to it. 

Reviewer: ```suggestion
meaning the template will show in your WordPress editor, but you don't have to stick to it. 
```
